Lib/xml/etree/ElementTree.py

```diff
@@ -1239,18 +1241,23 @@ def iterator(source):
                 pullparser.feed(data)
             root = pullparser._close_and_return_root()
             yield from pullparser.read_events()
-            it.root = root
+            it = wr()
+            if it is not None:
+                it.root = root
         finally:
             if close_source:
                 source.close()
 
     class IterParseIterator(collections.abc.Iterator):
         __next__ = iterator(source).__next__
-    it = IterParseIterator()
-    it.root = None
-    del iterator, IterParseIterator
 
-    next(it)
+        def __del__(self):
+            if close_source:
+                source.close()
+
+    it = IterParseIterator()
+    wr = weakref.ref(it)
+    del IterParseIterator
```

Comment:
I am curious, why previously both `iterator` and `IterParseIterator` names were deleted, but now only `IterParseIterator`? And what is the purpose of this statement in the first place? I was thinking that `iterator.__closure__` stores references to these objects; therefore, unnecessary references should be deleted. However, as per my checks, closure stores only referenced variables inside; `pullparser`, `close_source` and `wr` in this case.

Reply:
I also noticed that `it.root = None` was deleted. This fact is not [documented](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.iterparse), but this may still cause unintended errors on the user side if they use `root`.

Reply:
I think you are right about the `it.root = None`. I did not intend a behavioral change here, so it seems like a good idea to add it back.

I don't think the `del` statements matter one way or the other. They look like they break a cycle, but not really, but they also are harmless.